### PR TITLE
REF: BIP-47, improvements under the hood

### PR DESCRIPTION
**Pull request by Overtorment**

### CODE REVIEW

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/kalj2xkcoc4rcnmaf7voi3kzaq?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/1310)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/4ukakot5abvlwm7xly2dtrizzq?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/1311)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/zeztqhp44kkj3yzgly42ehpfda?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/1312)

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

File: class/wallets/abstract-hd-electrum-wallet.ts [hunk 1]
```diff
@@ -1495,6 +1501,100 @@ export class AbstractHDElectrumWallet extends AbstractHDWallet {
...
+
+    const targetsTemp: CreateTransactionTarget[] = [];
+    targetsTemp.push({
+      address: bobBip47.getNotificationAddress(),
+      value: 546, // minimum permissible utxo size
+    });
+    targetsTemp.push({
+      value: 0,
+      script: {
+        hex: '6a4c50' + blindedPaymentCodeTemp, // no `address` here, its gonabe op_return
```
Comment (limpbrains) [inline class/wallets/abstract-hd-electrum-wallet.ts:1538]: what is this hex value, comment would be nice

File: class/wallets/abstract-hd-electrum-wallet.ts [hunk 2]
```diff
@@ -1495,6 +1501,100 @@ export class AbstractHDElectrumWallet extends AbstractHDWallet {
...
+    const { inputs: inputsTemp } = this.createTransaction(
+      utxos,
+      targetsTemp,
+      feeRate,
+      changeAddress,
+      AbstractHDElectrumWallet.defaultRBFSequence,
+      false,
+      0,
+    );
+    assert(inputsTemp && inputsTemp[0] && inputsTemp[0].wif);
```
Comment (limpbrains) [inline class/wallets/abstract-hd-electrum-wallet.ts:1553]: inputsTemp?.[0]?.wif

File: class/wallets/abstract-hd-electrum-wallet.ts [hunk 3]
```diff
@@ -1495,6 +1501,100 @@ export class AbstractHDElectrumWallet extends AbstractHDWallet {
...
+
+    // utxo selected. lets recreate op_return payload using the correct (first!) utxo and correct targets with that payload
+
+    const keyPair = ECPair.fromWIF(inputsTemp[0].wif);
+    const outputNumber = Buffer.from('00000000', 'hex');
+    outputNumber.writeUInt32LE(+inputsTemp[0].vout);
```
Comment (limpbrains) [inline class/wallets/abstract-hd-electrum-wallet.ts:1559]: this should be already a number, no?
Why force?

File: class/wallets/abstract-hd-electrum-wallet.ts [hunk 4]
```diff
@@ -1495,6 +1501,100 @@ export class AbstractHDElectrumWallet extends AbstractHDWallet {
...
+
+    const targets: CreateTransactionTarget[] = [];
+    targets.push({
+      address: bobBip47.getNotificationAddress(),
+      value: 546, // minimum permissible utxo size
+    });
+    targets.push({
+      value: 0,
+      script: {
+        hex: '6a4c50' + blindedPaymentCode, // no `address` here, its gonabe op_return
```
Comment (limpbrains) [inline class/wallets/abstract-hd-electrum-wallet.ts:1577]: Probably should be a constant

File: class/wallets/abstract-hd-electrum-wallet.ts [hunk 5]
```diff
@@ -1536,6 +1636,31 @@ export class AbstractHDElectrumWallet extends AbstractHDWallet {
...
+    this._next_free_payment_code_address_index_receiver[receiverPaymentCode] =
+      this._next_free_payment_code_address_index_receiver[receiverPaymentCode] || 0; // init
+
+    for (let c = this._next_free_payment_code_address_index_receiver[receiverPaymentCode]; c < 999999; c++) {
+      const address = this._getBIP47AddressForReceiver(receiverPaymentCode, c);
+
+      this._addresses_by_payment_code_receiver[receiverPaymentCode] = this._addresses_by_payment_code_receiver[receiverPaymentCode] || {}; // init
+      this._addresses_by_payment_code_receiver[receiverPaymentCode][c] = address;
+      const histories = await BlueElectrum.multiGetHistoryByAddress([address]);
+      if (histories[address] && histories[address].length > 0) {
```
Comment (limpbrains) [inline class/wallets/abstract-hd-electrum-wallet.ts:1654]: Embrace optional chaining

File: class/wallets/abstract-hd-electrum-wallet.ts [hunk 6]
```diff
@@ -16,6 +16,7 @@ import ecc from '../../blue_modules/noble_ecc';
 import { randomBytes } from '../rng';
 import { AbstractHDWallet } from './abstract-hd-wallet';
 import { CreateTransactionResult, CreateTransactionTarget, CreateTransactionUtxo, Transaction, Utxo } from './types';
+import assert from 'assert';
```
Comment (limpbrains) [inline class/wallets/abstract-hd-electrum-wallet.ts:19]: I like assert

**Review by limpbrains — CHANGES_REQUESTED** [review]

File: class/wallets/abstract-hd-electrum-wallet.ts (hunk 1 above)
Comment (Overtorment) [reply class/wallets/abstract-hd-electrum-wallet.ts:1538]: ah yes, its literally OP_RETURN oppcode

**Comment by Overtorment** [comment]
@abhiShandy

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/6atljsmklkydqg5klczxsogf5i?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/1337)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/ia4avjwd6u2lfapj6xhsyh3wuy?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/1338)

File: class/wallets/abstract-hd-electrum-wallet.ts [hunk 7]
```diff
@@ -1495,6 +1501,87 @@ export class AbstractHDElectrumWallet extends AbstractHDWallet {
...
   }
 
+  async createBip47NotificationTransaction(
+    utxos: CreateTransactionUtxo[],
+    receiverPaymentCode: string,
+    feeRate: number,
+    changeAddress: string,
+  ) {
+    const aliceBip47 = BIP47Factory(ecc).fromBip39Seed(this.getSecret(), undefined, this.getPassphrase());
+    const bobBip47 = BIP47Factory(ecc).fromPaymentCode(receiverPaymentCode);
```
Comment (abhiShandy) [inline class/wallets/abstract-hd-electrum-wallet.ts:1511]: Can we validate the paymentCode string before using it?

**Review by abhiShandy — COMMENTED** [review]
I have left a few suggestions.

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/4rupzjbv3vtpt7bpzeeqcm52sq?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/1341)

File: class/wallets/abstract-hd-electrum-wallet.ts (hunk 7 above)
Comment (Overtorment) [reply class/wallets/abstract-hd-electrum-wallet.ts:1511]: checksum is validated somewhere indise, it throws an exception if its not a valid PC

Comment (abhiShandy) [reply class/wallets/abstract-hd-electrum-wallet.ts:1537]: alright. Do we need to unit test that?

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/fkvu6hti3ghtez7skfs4rbpw2a?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/1397)

**Pull request merged**